src/Compilers/CSharp/Test/Semantic/Semantics/LambdaTests.cs

```diff
@@ -3480,5 +3486,745 @@ void F()
+                parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+            var exprs = tree.GetRoot().DescendantNodes().OfType<LambdaExpressionSyntax>().ToImmutableArray();
+            Assert.Equal(3, exprs.Length);
+            var lambdas = exprs.SelectAsArray(e => GetLambdaSymbol(model, e));
+            Assert.Equal(FlowAnalysisAnnotations.MaybeNull | FlowAnalysisAnnotations.NotNull, lambdas[0].ReturnTypeFlowAnalysisAnnotations);
+            Assert.Equal(new[] { "obj" }, lambdas[1].ReturnNotNullIfParameterNotNull);
+            Assert.Equal(new[] { "x" }, lambdas[2].NotNullMembers);
+            Assert.Equal(new[] { "y" }, lambdas[2].NotNullWhenFalseMembers);
+            Assert.Equal(new[] { "z" }, lambdas[2].NotNullWhenTrueMembers);
+        }
+
+        [Fact]
+        public void LambdaAttributes_NullableAttributes_02()
+        {
+            var source =
+@"#nullable enable
+using System;
+using System.Diagnostics.CodeAnalysis;
+class Program
+{
+    static void Main()
+    {
+        Func<object> a1 = [return: MaybeNull] () => null;
+        Func<object?> a2 = [return: NotNull] () => null;
+    }
+}";
+            var comp = CreateCompilation(new[] { source, MaybeNullAttributeDefinition, NotNullAttributeDefinition }, parseOptions: TestOptions.RegularPreview);
+            // PROTOTYPE: Report error for a2, no error for a1.
+            comp.VerifyDiagnostics(
+                // (8,53): warning CS8603: Possible null reference return.
+                //         Func<object> a1 = [return: MaybeNull] () => null;
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "null").WithLocation(8, 53));
+        }
+
+        [Fact]
+        public void LambdaAttributes_NullableAttributes_03()
+        {
+            var source =
+@"#nullable enable
+using System;
+using System.Diagnostics.CodeAnalysis;
+class Program
+{
+    static void Main()
+    {
+        Action<object> a1 = ([AllowNull] x) => { x.ToString(); };
+        Action<object?> a2 = ([DisallowNull] x) => { x.ToString(); };
+    }
+}";
+            var comp = CreateCompilation(new[] { source, AllowNullAttributeDefinition, DisallowNullAttributeDefinition }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (8,50): warning CS8602: Dereference of a possibly null reference.
+                //         Action<object> a1 = ([AllowNull] x) => { x.ToString(); };
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(8, 50));
+        }
+
+        [Fact]
+        public void LambdaAttributes_DoesNotReturn()
+        {
+            var source =
+@"using System;
+using System.Diagnostics.CodeAnalysis;
+class Program
+{
+    static void Main()
+    {
+        Action a1 = [DoesNotReturn] () => { };
+        Action a2 = [DoesNotReturn] () => throw new Exception();
```

Comment:
Please add a case where the lambda does return (we'll enforce/warn)

Reply:
Should warn on `a1` too

---
In reply to: [612139330](https://github.com/dotnet/roslyn/pull/52178#discussion_r612139330) [](ancestors = 612139330)

Reply:
From offline discussion, let's test NotNullWhen enforcement too (should likely fall out when DoesNotReturn works)

---
In reply to: [612140847](https://github.com/dotnet/roslyn/pull/52178#discussion_r612140847) [](ancestors = 612140847,612139330)